src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +432,545 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
   · simp [of_length_zero]
   · simp [shiftLeftRec_eq]
 
+/-! # udiv/urem recurrence for bitblasting
+
+Let us study an instructive counterexample to the claim that
+  `n = d * q + r` for (`0 ≤ r < d`) uniquely determining q and r *over bitvectors*.
+
+- Let `bitwidth = 3`
+- Let `n = 0, d = 3`
+- If we choose `q = 2, r = 2`, then d * q + r = 6 + 2 = 8 ≃ 0 (mod 8) so satisfies.
+- But see that `q = 0, r = 0` also satisfies, as 0 * 3 + 0 = 0.
+- So for (`n = 0, d = 3`), both:
+    `q = 2, r = 2` as well as
+    `q = 0, r = 0` are solutions!
+
+It's easy to cook up such examples, by chosing `(q, r)` for a fixed `(d, n)`
+such that `(d * q + r)` overflows.
+
+This tells us that the division algorithm must have more restrictions that just the ones
+we have for natural numbers. These restrictions are captured in `DivRemInput.Lawful`,
+which captures the relationship necessary between `n, d, q, r`. The key idea is to state
+the relationship in terms of the `{n, d, q, r}.toNat` values, and then prove that the relationship
+holds for the bitvector values.
+
+References:
+- Fast 32-bit Division on the DSP56800E: Minimized nonrestoring division algorithm by David Baca
+- Bitwuzla sources for bitblasting.h
+-/
+
+
+/-- TODO: This theorem surely exists somewhere, but I can't find it. -/
```

Comment:
This needs to move into its place. And you can probably drop this comment if you cannot find it.

Reply:
@tobiasgrosser It feels a bit too special case to move into `Nat`, but I'm happy to move it...

Reply:
Then just drop the comment.

Reply:
Also, there are too many linebreaks before this theorem.